File: artifacts/api-server/src/lib/providers/easemytrip.ts
import type { MappedStatus } from "../train-status-mapper";
import { TrainStatusNotFoundError, TrainStatusUpstreamError } from "./errors";
import { fetchProviderStatus } from "./http";
import {
  assembleMappedStatus,
  type AssembleOptions,
  type ProviderStationRow,
} from "./normalize";
import type { KnownTrain, ProviderFetchOptions, TrainStatusProvider } from "./types";

const EASEMYTRIP_ENDPOINT =
  "https://www.easemytrip.com/railways/train-live-status-with-click";

const HEADERS: Record<string, string> = {
  "User-Agent":
    "Mozilla/5.0 (compatible; TrainTracker/1.0) AppleWebKit/537.36",
  Accept: "text/html,application/xhtml+xml",
};

/** YYYYMMDD -> DD/MM/YYYY, or null when not a valid 8-digit date. */
export function toEaseMyTripDate(departureDate: string): string | null {
  if (!/^\d{8}$/.test(departureDate)) return null;
  return `${departureDate.slice(6, 8)}/${departureDate.slice(4, 6)}/${departureDate.slice(0, 4)}`;
}

const STATION_BLOCK_OPEN = '<div class="station-item ">';
const CODE_RE = /<div class="code">([^<]*)<\/div>/;
const KM_RE = /<div class="km">([^<]*)<\/div>/;
const CURRENT_DOT_RE = /class="dot[^"]*blink blue/;
const STATION_NAME_RE = /<div class="[^"]*station-name[^"]*">([\s\S]*?)<\/div>/;
const STRONG_NAME_RE = /<strong>([^<]*)<\/strong>/;
const ANCHOR_NAME_RE = /<a[^>]*>\s*<span>([^<]*)<\/span>/;
const SPAN_NAME_RE = /<span>([^<]*)<\/span>/;
const PLATFORM_RE = /PF\s*([^<\s]*)</;
const DELAY_RE = /<div class="delay">([\s\S]*?)<\/div>/;
const ARRIVAL_RE = /<div class="avl_times">[\s\S]*?<span>([^<]*)<\/span>/;
const DEPARTURE_RE = /<div class="dprt_times">[\s\S]*?<span>([^<]*)<\/span>/;
const TITLE_RE = /<h2[^>]*>\s*(\d{4,5})\s+([\s\S]*?)Running Status/;
const LAST_UPDATED_RE = /<strong>Last Updated:<\/strong>\s*([\s\S]*?)<\/p>/;

function stripTags(html: string): string {
  return html.replace(/<[^>]*>/g, "").trim();
}

interface ParsedStation {
  code: string;
  name: string;
  platform: string | null;
  delay_minutes: number | null;
  arrival: string | null;
  departure: string | null;
  distance: number | null;
  is_current: boolean;
}

function parseDelay(inner: string): number | null {
  if (/On Time/i.test(inner)) return 0;
  const late = /Late by\s*(\d+)\s*min/i.exec(inner);
  if (late) return Number(late[1]);
  const early = /Early by\s*(\d+)\s*min/i.exec(inner);
  if (early) return -Number(early[1]);
  return null;
}

function parseTime(inner: string | null): string | null {
  const time = inner?.trim() ?? "";
  return time === "" ? null : time;
}

function parseStationBlock(block: string): ParsedStation | null {
  const codeMatch = CODE_RE.exec(block);
  if (!codeMatch) return null;

  const nameMatch = STATION_NAME_RE.exec(block);
  let name = "";
  let platform: string | null = null;
  if (nameMatch) {
    const inner = nameMatch[1];
    name =
      STRONG_NAME_RE.exec(inner)?.[1] ??
      ANCHOR_NAME_RE.exec(inner)?.[1] ??
      SPAN_NAME_RE.exec(inner)?.[1] ??
      stripTags(inner);
    const pf = PLATFORM_RE.exec(inner);
    platform = pf && pf[1] ? pf[1].trim() : null;
  }

  const delayMatch = DELAY_RE.exec(block);
  const arrivalMatch = ARRIVAL_RE.exec(block);
  const departureMatch = DEPARTURE_RE.exec(block);
  const kmMatch = KM_RE.exec(block);
  const km = kmMatch ? Number(kmMatch[1].replace("Km", "").trim()) : NaN;

  return {
    code: codeMatch[1].trim(),
    name: name.trim(),
    platform,
    delay_minutes: delayMatch ? parseDelay(delayMatch[1]) : null,
    arrival: parseTime(arrivalMatch?.[1] ?? null),
    departure: parseTime(departureMatch?.[1] ?? null),
    distance: Number.isFinite(km) ? km : null,
    is_current: CURRENT_DOT_RE.test(block),
  };
}

/** Parse the EaseMyTrip running-status page into normalized rows. */
export function parseEaseMyTripHtml(html: string): {
  rows: ProviderStationRow[];
  currentStationCode: string | null;
  titleTrainNumber: string | null;
  lastUpdated: string | null;
} {
  const title = TITLE_RE.exec(html);
  const titleTrainNumber = title ? title[1] : null;

  const lastUpdatedMatch = LAST_UPDATED_RE.exec(html);
  const lastUpdated = lastUpdatedMatch
    ? stripTags(lastUpdatedMatch[1])
    : null;

  const blocks = html.split(STATION_BLOCK_OPEN);
  let currentStationCode: string | null = null;
  const rows: ProviderStationRow[] = [];
  for (const block of blocks) {
    const parsed = parseStationBlock(block);
    if (!parsed) continue;
    if (parsed.is_current) currentStationCode = parsed.code;
    rows.push({
      station_code: parsed.code,
      station_name: parsed.name,
      scheduled_arrival: parsed.arrival,
      actual_arrival: null,
      scheduled_departure: parsed.departure,
      actual_departure: null,
      delay_minutes: parsed.delay_minutes,
      distance: parsed.distance,
      platform: parsed.platform,
      halt_minutes: null,
      day: 1,
    });
  }

  return { rows, currentStationCode, titleTrainNumber, lastUpdated };
}

/** Map the EaseMyTrip page to a `MappedStatus`. */
export function mapEaseMyTripHtml(
  html: string,
  options: AssembleOptions,
): MappedStatus {
  const parsed = parseEaseMyTripHtml(html);

  if (parsed.rows.length === 0) {
    // A page that does not even name the requested train means "not found".
    // If the train name is present but no stations parsed, the layout likely
    // changed — surface as an upstream error rather than cache a false 404.
    if (parsed.titleTrainNumber === null) {
      throw new TrainStatusNotFoundError(
        "easemytrip",
        "Train not found or no data",
      );
    }
    throw new TrainStatusUpstreamError(
      "easemytrip",
      "No stations parsed from EaseMyTrip page",
    );
  }

  return assembleMappedStatus(parsed.rows, {
    ...options,
    currentStationCode: parsed.currentStationCode,
    statusMessage: null,
    lastUpdated: parsed.lastUpdated,
  });
}

export class EaseMyTripProvider implements TrainStatusProvider {
  readonly name = "easemytrip";
  readonly enabled = true;

  fetchTrainStatus(
    trainNumber: string,
    departureDate: string,
    options: ProviderFetchOptions,
    knownTrain: KnownTrain | null = null,
  ): Promise<MappedStatus> {
    const emtDate = toEaseMyTripDate(departureDate);
    if (!emtDate) {
      return Promise.reject(
        new TrainStatusNotFoundError(
          this.name,
          `Unsupported date format: ${departureDate}`,
        ),
      );
    }

    const url = new URL(EASEMYTRIP_ENDPOINT);
    url.searchParams.set("trainnumber", trainNumber);
    url.searchParams.set("date", emtDate);

    return fetchProviderStatus(
      {
        provider: this.name,
        url: url.toString(),
        responseType: "text",
        headers: HEADERS,
        map: (html) =>
          mapEaseMyTripHtml(String(html), {
            trainNumber,
            departureDate,
            knownTrain,
          }),
      },
      options,
    );
  }
}

export function createEaseMyTripProvider(): TrainStatusProvider {
  return new EaseMyTripProvider();
}

File: rust/crates/provider-easemytrip/tests/easemytrip.rs
//! Port of `lib/providers/easemytrip.test.ts`, with the fetch leg driven
//! through [`MockTransport`].

use std::sync::Arc;

use tt_mapper::{AssembleOptions, KnownTrain};
use tt_provider_core::{ProviderError, ProviderFetchOptions, TrainStatusProvider};
use tt_provider_easemytrip::{
    create_easemytrip_provider, map_easemytrip_html, parse_easemytrip_html, to_easemytrip_date,
    EaseMyTripProvider,
};
use tt_provider_http::MockTransport;

const ENDPOINT_PATH: &str = "/railways/train-live-status-with-click";

#[derive(Default)]
struct BlockOptions {
    current: bool,
    km: Option<&'static str>,
    delay: Option<&'static str>,
    arrival: Option<&'static str>,
    departure: Option<&'static str>,
    platform: Option<&'static str>,
}

fn station_block(code: &str, name: &str, opts: BlockOptions) -> String {
    let current_col = if opts.current { "arvd_sts_col" } else { " " };
    let current_dot = if opts.current {
        "arvd_sts_col_dot blink blue"
    } else {
        ""
    };
    let name_class = if opts.current {
        "arvd_sts_sec station-name"
    } else {
        "station-name"
    };
    format!(
        r#"<div class="station-item ">
    <div class="station-item_inner">
        <div class="left-col">
            <div class="code">{code}</div>
            <div class="km">{km}</div>
        </div>
        <div class="line-col {current_col}">
            <div class="dot {current_dot}"></div>
        </div>
        <div class="right-col">
            <div class="{name_class}">
                <a href="https://www.easemytrip.com/railways/x/">
                    <span>{name}</span>
                </a>
                <span>PF {platform}</span>
            </div>
            <div class="delay">
                <span>{delay}</span>
            </div>
            <div class="avl_times">
                <span>{arrival}</span>
            </div>
            <div class="dprt_times">
                <span>{departure}</span>
                <span class=""></span>
            </div>
        </div>
    </div>
</div>"#,
        km = opts.km.unwrap_or("0.0 Km"),
        delay = opts.delay.unwrap_or("On Time"),
        arrival = opts.arrival.unwrap_or(""),
        departure = opts.departure.unwrap_or(""),
        platform = opts.platform.unwrap_or(""),
    )
}

const INTERMEDIATE_BLOCK: &str = r#"<div class="station-item ">
    <div class="station-item_inner inner_station_btn_sec">
        <p class="inner_station_btn">17 Intermediate Stations</p>
    </div>
    <div id="intermediate_0" style="display:none;">
        <div class="station-item_inner inner_station_div">
            <div class="station-name">
                <span>Bby Mahalakshmi</span>
                <span>PF </span>
            </div>
            <div class="delay"><span>-</span></div>
            <div class="avl_times"><span>17:01</span></div>
            <div class="dprt_times"><span>17:01</span></div>
        </div>
    </div>
</div>"#;

const HEADER_BLOCK: &str = r#"<div class="station-item station-item_top blu">
    <div class="station-item_inner">
        <div class="station-name">Stations</div>
    </div>
</div>
<div class="station-item date_label_prnt">
    <div class="station-item_inner ">Day 1</div>
</div>"#;

fn fixture_html() -> String {
    format!(
        r#"<html><body>
    <h2 class="bs-pra mf16 mt10">
        12951 Tejas Rajdhani Running Status
    </h2>
    <p class="f11_ mt_10">
        <strong>Last Updated:</strong>
        10 minutes ago
    </p>
    {header}
    {mmct}
    {intermediate}
    {bvi}
    {ndls}
</body></html>"#,
        header = HEADER_BLOCK,
        mmct = station_block(
            "MMCT",
            "Mumbai Central",
            BlockOptions {
                current: true,
                departure: Some("17:00"),
                ..BlockOptions::default()
            },
        ),
        intermediate = INTERMEDIATE_BLOCK,
        bvi = station_block(
            "BVI",
            "Borivali",
            BlockOptions {
                km: Some("30.0 Km"),
                arrival: Some("17:20"),
                departure: Some("17:22"),
                platform: Some("6"),
                ..BlockOptions::default()
            },
        ),
        ndls = station_block(
            "NDLS",
            "New Delhi",
            BlockOptions {
                km: Some("1384.0 Km"),
                delay: Some("Late by 10 min"),
                arrival: Some("08:32"),
                platform: Some("3"),
                ..BlockOptions::default()
            },
        ),
    )
}

fn known_train() -> KnownTrain {
    KnownTrain {
        number: "12951".to_string(),
        name: "Tejas Rajdhani Express".to_string(),
    }
}

fn assemble_options() -> AssembleOptions {
    AssembleOptions {
        train_number: "12951".to_string(),
        departure_date: "20260806".to_string(),
        known_train: Some(known_train()),
        ..AssembleOptions::default()
    }
}

#[test]
fn to_easemytrip_date_converts_yyyymmdd_to_dd_mm_yyyy() {
    assert_eq!(
        to_easemytrip_date("20260806"),
        Some("06/08/2026".to_string())
    );
}

#[test]
fn to_easemytrip_date_returns_none_for_malformed_dates() {
    assert_eq!(to_easemytrip_date("bogus"), None);
}

#[test]
fn parses_real_stations_and_skips_intermediate_stations() {
    let parsed = parse_easemytrip_html(&fixture_html());

    assert_eq!(parsed.title_train_number.as_deref(), Some("12951"));
    assert_eq!(parsed.last_updated.as_deref(), Some("10 minutes ago"));
    assert_eq!(parsed.current_station_code.as_deref(), Some("MMCT"));
    let codes: Vec<_> = parsed
        .rows
        .iter()
        .map(|row| row.station_code.as_str())
        .collect();
    assert_eq!(codes, ["MMCT", "BVI", "NDLS"]);

    let mmct = &parsed.rows[0];
    assert_eq!(mmct.station_code, "MMCT");
    assert_eq!(mmct.station_name, "Mumbai Central");
    assert_eq!(mmct.scheduled_departure.as_deref(), Some("17:00"));
    assert_eq!(mmct.scheduled_arrival, None);
    assert_eq!(mmct.delay_minutes, Some(0));
    assert_eq!(mmct.distance, Some(0));
    assert_eq!(mmct.platform, None);

    let bvi = &parsed.rows[1];
    assert_eq!(bvi.station_code, "BVI");
    assert_eq!(bvi.station_name, "Borivali");
    assert_eq!(bvi.scheduled_arrival.as_deref(), Some("17:20"));
    assert_eq!(bvi.scheduled_departure.as_deref(), Some("17:22"));
    assert_eq!(bvi.platform.as_deref(), Some("6"));
    assert_eq!(bvi.distance, Some(30));
    assert_eq!(bvi.delay_minutes, Some(0));

    let ndls = &parsed.rows[2];
    assert_eq!(ndls.station_code, "NDLS");
    assert_eq!(ndls.station_name, "New Delhi");
    assert_eq!(ndls.scheduled_arrival.as_deref(), Some("08:32"));
    assert_eq!(ndls.scheduled_departure, None);
    assert_eq!(ndls.platform.as_deref(), Some("3"));
    assert_eq!(ndls.delay_minutes, Some(10));
}

#[test]
fn maps_the_page_to_a_full_status() {
    let status = map_easemytrip_html(&fixture_html(), &assemble_options()).expect("maps");

    assert_eq!(status.train_name, "Tejas Rajdhani Express");
    assert_eq!(status.source_station_code, "MMCT");
    assert_eq!(status.destination_station_code, "NDLS");
    assert_eq!(status.current_station_code.as_deref(), Some("MMCT"));
    assert_eq!(
        status.current_station_name.as_deref(),
        Some("Mumbai Central")
    );
    assert_eq!(status.current_delay_minutes, Some(0));
    assert_eq!(status.last_updated.as_deref(), Some("10 minutes ago"));
    assert_eq!(status.status_message, None);

    assert!(status.stations[0].is_current);
    assert!(!status.stations[0].has_departed);
    assert!(!status.stations[1].has_departed);
    assert!(!status.stations[2].has_departed);
}

#[test]
fn reports_not_found_when_the_page_does_not_name_the_train() {
    let err = map_easemytrip_html("<html><body><p>oops</p></body></html>", &assemble_options())
        .expect_err("must be not-found");
    assert!(matches!(err, ProviderError::NotFound { .. }));
}

#[test]
fn reports_upstream_error_when_title_exists_but_no_stations_parse() {
    let err = map_easemytrip_html(
        "<h2>12951 Tejas Rajdhani Running Status</h2><p>layout changed</p>",
        &assemble_options(),
    )
    .expect_err("must be upstream");
    assert!(matches!(err, ProviderError::Upstream { .. }));
}

#[test]
fn builds_the_expected_url_and_returns_mapped_status() {
    let mut transport = MockTransport::new();
    transport.push(ENDPOINT_PATH, 200, fixture_html().into_bytes());
    let transport = Arc::new(transport);
    let provider = EaseMyTripProvider::new(transport.clone());

    let status = tokio::runtime::Runtime::new()
        .unwrap()
        .block_on(provider.fetch_train_status(
            "12951",
            "20260806",
            &ProviderFetchOptions::default(),
            Some(&known_train()),
        ))
        .expect("fetch");

    assert_eq!(status.current_station_code.as_deref(), Some("MMCT"));

    let url = url::Url::parse(&transport.requests()[0].url).expect("valid url");
    assert_eq!(url.scheme(), "https");
    assert_eq!(url.host_str(), Some("www.easemytrip.com"));
    assert_eq!(url.path(), ENDPOINT_PATH);
    let pairs: Vec<_> = url.query_pairs().collect();
    assert!(pairs.contains(&("trainnumber".into(), "12951".into())));
    assert!(pairs.contains(&("date".into(), "06/08/2026".into())));
}

#[test]
fn factory_returns_the_provider_with_the_right_name() {
    let provider = create_easemytrip_provider(Arc::new(MockTransport::new()));
    assert_eq!(provider.name(), "easemytrip");
}
